### PR TITLE
Allow transferring position with an optional external notification callback

### PR DESCRIPTION
**Pull request by lbeder**
This PR adds the `transferPositionAndCall` function, allowing the caller to specify a custom transfer position notification callback.

### CODE REVIEW

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol [hunk 1]
```diff
@@ -763,8 +764,53 @@ contract LiquidityProtection is ILiquidityProtection, Utils, Owned, ReentrancyGu
...
+    function transferPositionAndCall(
+        uint256 id,
+        address newProvider,
+        address target,
+        bytes memory data
+    ) external protected validAddress(newProvider) returns (uint256) {
+        uint256 newId = transferPosition(msg.sender, id, newProvider);
+
+        // make sure that we're not trying to call into the zero address or a fallback function
+        require(target != address(0) && data.length > 0, "ERR_INVALID_CALL_DATA");
```
Comment (barakman) [inline solidity/contracts/liquidity-protection/LiquidityProtection.sol:789]: Alternative 1:
```
        validAddress(target);
        ...
        require(data.length > 0, "ERR_INVALID_CALL_DATA");
```

Alternative 2:
```
        require(target != address(0), "ERR_INVALID_CALL_TARGET");
        require(data.length > 0, "ERR_INVALID_CALL_DATA");
```

Comment (barakman) [reply solidity/contracts/liquidity-protection/LiquidityProtection.sol:789]: Also, it's possibly better to check `data.length >= 4` (length of function selector), but I'm not sure about this.

**Review by barakman — CHANGES_REQUESTED** [review]

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol (hunk 1 above)
Comment (lbeder) [reply solidity/contracts/liquidity-protection/LiquidityProtection.sol:789]: Regarding the alternatives, I'm not sure there's any benefit in separating this test into two tests. I will extend the `data.length`, but keep in mind that we might need to remove it entirely if we'll decide to support fallback functions.

**Review by barakman — APPROVED** [review]

**Review by barakman — APPROVED** [review]

File: solidity/contracts/liquidity-protection/LiquidityProtection.sol [hunk 2]
```diff
@@ -763,8 +765,53 @@ contract LiquidityProtection is ILiquidityProtection, Utils, Owned, ReentrancyGu
...
+    function transferPositionAndCall(
+        uint256 id,
+        address newProvider,
+        address target,
+        bytes memory data
+    ) external protected validAddress(newProvider) returns (uint256) {
+        uint256 newId = transferPosition(msg.sender, id, newProvider);
+
+        // make sure that we're not trying to call into the zero address or a fallback function
+        require(target != address(0) && data.length >= FUNC_SELECTOR_LENGTH, "ERR_INVALID_CALL_DATA");
```
Comment (yudilevi) [inline solidity/contracts/liquidity-protection/LiquidityProtection.sol:790]: any reason not to call `validAddress` on the target instead?

Comment (lbeder) [reply solidity/contracts/liquidity-protection/LiquidityProtection.sol:790]: @barakman mentioned it as well, so I'll just fix it 😎

**Review by yudilevi — APPROVED** [review]

**Pull request merged**